contracts/transfer/src/state.rs

```diff
@@ -198,45 +170,54 @@ impl TransferState {
         &mut self,
         tx: Transaction,
     ) -> Result<Vec<u8>, ContractError> {
-        //  1. α ∈ R
-        if !self.root_exists(&tx.anchor) {
-            panic!("Anchor not found in the state!");
+        let tx_skeleton = tx.payload().tx_skeleton();
+
+        // panic if the transaction fails basic validation
+        assert!(tx.is_valid(), "Transaction is invalid");
```

Comment:
is_valid() checks the payload hash correctness, several remarks to that:
- is this new, I think it is, if so, why was it introduced?
- this duplicates the checks done in preverify and elsewhere, why doing it here? This would be yet another hash calculation which is wasteful unless there is a good reason for it (I am not saying, just asking)
- for deployment transaction this will be a problem, as it eliminates the possibility of editing the transaction before passing it here (for deployment transaction we will need such edit as we want to replace bytecode with the hash of bytecode)

Reply:
> this duplicates the checks done in preverify

The single source of true is the contract
The contract MUST do all the checks, so IMO this is correct

Ps: you can't rely on checks being done by preverify, because preverify is not run while executing blocks

Reply:
> it eliminates the possibility of editing the transaction before passing it

Why do we need this? Transactions SHALL NOT be edited, otherwise their ID will change with unforeseen side effects

Reply:
> transaction did not have a payload_hash equivalent

Yes they have, their ID is the hash

Reply:
if ID is the hash, then why do we need field 'payload_hash' ?

Reply:
`Transactions SHALL NOT be edited, otherwise their ID will change with unforeseen side effects` - that sounds really good to me and I fully agree - I was basing my reasoning on the statement by, I think it was Ed, who said that we need to convert bytecode to hash of bytecode, because bytecode should not enter the blockchain. This is why I was talking about editing of the transaction, although I agree that it should not be edited

Reply:
Re: contract being source of truth - that is fine, yet we do not need duplicate checks, if circuit checks something, contract does not have to - that was my point.

Reply:
The `payload_hash` is not new but has another name after changes to the phoenix paper. Previously it was known as the transaction hash and it is a vital part of the phoenix proof.

The main reason of this hash is precisely that we *don't* want the content of a transaction to change after it has been submitted. To ensure this, the transaction is hashed and signed and the signature will be proven in circuit.

That being said, we *do* have some freedom in how we generate the input to the hash of a transaction, trough the method `Transaction::to_hash_input_bytes`.
So with your above problem with the deployment byte-code being replaced by the hash, I think we might be able to generate the hash-input from the `ContractCall` by already using the hash of the bytecode. This way we don't change the `payload_hash` when the hash replaces the byte-code later.

Reply:
And with regards to the thought behind the `is_valid` method:
My idea was to detect obviously bogus `Transaction`s as early as possible and not only in the computationally very heavy phase of proof-creation. It is a very basic check but it makes it harder to flood the network with fake transactions.

Reply:
We agreed to remove the `payload_hash` field from the `Transaction` struct.
This is *not* so that the `Transaction` (or more precise: the preimage of the hash) can be changed after it has been created, but rather so that the verifier is forced to newly compute the `payload_hash` for the proof the verification.